File: typescript-examples/src/module1/testDoubles/spy/UserService.ts
import {EmailSender} from "./EmailSender.ts";

export class UserService {
    private readonly emailSender: EmailSender;

    constructor(emailSender: EmailSender) {
        this.emailSender = emailSender;
    }

    public registerUser(email: string): void {
        // Register user logic
        this.emailSender.sendEmail(email, "Welcome!");
    }
}

File: module-1-content/src/main/java/com/codurance/module1/TestDoubles/Spy/UserService.java
package com.codurance.module1.TestDoubles.Spy;

public class UserService {
    private EmailSender emailSender;

    public UserService(EmailSender emailService) {
        this.emailSender = emailService;
    }

    public void registerUser(String email) {
        // Register user logic
        emailSender.sendEmail(email, "Welcome!");
    }
}
